enclave/src/polkadex.rs

```diff
@@ -0,0 +1,151 @@
+    for account in accounts.iter() {
+        if account.account.prev == last_account.as_ref() {
+            StorageProofChecker::<<Header as HeaderT>::Hashing>::check_proof(
+                header.state_root,
+                account.account.current.as_ref(), // QUESTION: How is this key defined? What about storage prefix?
+                account.proof.to_vec(),
+            )
+            .sgx_error_with_log("Erroneous StorageProof")?;
+
+            last_account = AccountId::from(account.account.next.clone().unwrap());
+        }
+    }
+
+    Ok(())
+}
+
+pub fn create_in_memory_account_storage(accounts: Vec<PolkadexAccount>) -> SgxResult<()> {
+    let accounts_storage = PolkadexAccountsStorage::create(accounts);
+    let storage_ptr = Arc::new(SgxMutex::<PolkadexAccountsStorage>::new(accounts_storage));
+    let ptr = Arc::into_raw(storage_ptr);
+    GLOBAL_ACCOUNTS_STORAGE.store(ptr as *mut (), Ordering::SeqCst);
+    Ok(())
+}
+
+/// Access that pointer
+pub struct PolkadexAccountsStorage {
+    accounts: HashMap<[u8;32], Vec<[u8;32]>>,
+}
+
+impl PolkadexAccountsStorage {
+    pub fn create(accounts: Vec<PolkadexAccount>) -> PolkadexAccountsStorage {
+        let mut in_memory_map: PolkadexAccountsStorage = PolkadexAccountsStorage {
+            accounts: HashMap::new(),
+        };
+        for account in accounts {
+            in_memory_map
+                .accounts
+                .insert(account.account.current, account.account.proxies);
+        }
+        in_memory_map
+    }
+
+    pub fn add_main_account(&mut self, acc: [u8;32]) {
+        let vec: Vec<[u8;32]> = Vec::new();
+        self.accounts.insert(acc, vec);
+    }
+
+    pub fn remove_main_account(&mut self, acc: [u8;32]) {
+        self.accounts.remove(&acc);
+    }
+
+    pub fn add_proxy(&mut self, main: [u8;32], proxy: [u8;32]) {
+        let proxies: &mut Vec<[u8;32]> = self.accounts.get_mut(&main).unwrap();
+        proxies.push(proxy);
+    }
+
+    pub fn remove_proxy(&mut self, main: [u8;32], proxy: [u8;32]) {
+        let proxies: &mut Vec<[u8;32]> = self.accounts.get_mut(&main).unwrap();
+        for index in 0..proxies.len() {
+            if proxies.get(index).unwrap() == &proxy {
+                proxies.remove(index);
+                break;
+            }
+        }
+    }
+}
+
+pub fn check_main_account(acc: [u8;32]) -> SgxResult<bool> {
+    // Aquire lock on proxy_registry
+    let mutex = load_proxy_registry()?;
+    let mut proxy_storage: SgxMutexGuard<PolkadexAccountsStorage> = mutex.lock().unwrap();
+    Ok(proxy_storage.accounts.contains_key(&acc))
+}
+
+pub fn check_proxy_account(main_acc: [u8;32], proxy: [u8;32]) -> SgxResult<bool> {
+    // Aquire lock on proxy_registry
+    let mutex = load_proxy_registry()?;
+    let mut proxy_storage: SgxMutexGuard<PolkadexAccountsStorage> = mutex.lock().unwrap();
+
+    let list_of_proxies = proxy_storage.accounts.get(&main_acc).unwrap(); //FIXME: Remove Unwrap
```

Comment:
In case the main account is not registered, this function will panic. Please ensure that this will not happen (and if you already ensured that main account exists, then please add a short comment how/where it is done)